From 08ae173cc6cfa29e637444aae71b9e82716b0b82 Mon Sep 17 00:00:00 2001
From: Romain Dolbeau <romain@dolbeau.org>
Date: Sun, 25 Oct 2026 15:52:17 +0100
Subject: [PATCH 08/10] m68k pmap (MULTIPROCESSOR): clear the wired bit
 atomically in pmap_unwire

pmap_unwire() cleared PG_W with a plain read-modify-write of the PTE;
another CPU's MMU may set U or M between the read and the write, and a
lost M bit can lose the contents of a dirty page.  Use pmap_pte_update()
(compare-and-swap) like the other PTE updates of the MP pmap.
---
 sys/arch/m68k/m68k/pmap_motorola.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/sys/arch/m68k/m68k/pmap_motorola.c b/sys/arch/m68k/m68k/pmap_motorola.c
index 08e14b695..402f54de9 100644
--- a/sys/arch/m68k/m68k/pmap_motorola.c
+++ b/sys/arch/m68k/m68k/pmap_motorola.c
@@ -2481,7 +2481,12 @@ pmap_unwire(pmap_t pmap, vaddr_t va)
 	 */
 
 	if (pmap_pte_w_chg(pte, 0)) {
+#ifdef MULTIPROCESSOR
+		/* not a plain RMW: another CPU's MMU may set U or M meanwhile */
+		(void)pmap_pte_update(pte, 0, (pt_entry_t)~PG_W);
+#else
 		pmap_pte_set_w(pte, false);
+#endif
 		pmap->pm_stats.wired_count--;
 	}
 }
-- 
2.47.3

